### DISCUSSION TITLE
Extend API for video encoding

### DISCUSSION THREAD

**Discussion by fsiddi** [cgwire/kitsu#1243] [Ideas]
When submitting a video preview to Kitsu, the file is processed through [ffmpeg](https://github.com/cgwire/zou/blob/091d9190c0e129ca3520e05a3d23a31217906833/zou/utils/movie.py#L230). The current settings (especially the high def one) deliver often unnecessarily large files, taking a lot of compute time and storage.

I'd like to suggest the introduction of an option to skip the generation of such videos altogether, and rely on the submitted video instead. By providing TDs with the exact specs of the video you expect (especially keyframes, etc), the "time-to-preview" can be cut to zero.

I understand that the video should be following the specs, otherwise you can't guarantee good performance. This could be automated, using ffprobe to check if the video is compliant, and then kick off the re-encoding pipeline if needed.

Alternatively, an "encoding settings" page could be set up, where admins are allowed to set the bitrate and resolution for the re-encoding.

**Comment by EmberLightVFX** [comment]
> deliver often unnecessarily large files, taking a lot of compute time and storage.
This was changed not long time ago to make the scrubbing for the reviews way smoother by adding a lot of keyframes.

But yes, being able to playback the original file could be good, if the player supports it.

**Comment by fsiddi** [comment]
> This was changed not long time ago to make the scrubbing for the reviews way smoother by adding a lot of keyframes.

I suspect it's the `28M` that makes the files large.

**Comment by frankrousseau** [comment]
The problem is that normalizing videos makes things much more maintainable for us. We have two complex video players and a playlist-build system to deal with. That said, it's a hot topic, so we will think about it.

Notes: 
- You can already do it via the Gazu client
- Many studios want high-end videos for their review, they would take zero-loss videos if it was possible
- I add this issue as a reference. Topics are related: https://github.com/cgwire/kitsu/issues/1146

**Reply by fsiddi** [reply]
Thanks for the reply. I understand that there are several requirements in place. I think that allowing configuring the LD/HD encoding settings won't conflict with any requirements. I can at least provide a mockup if you are interested.

About your notes, I checked the docs https://gazu.cg-wire.com/examples#post-a-preview and https://api-docs.kitsu.cloud/operation/operation-post-pictures-preview-files-parameter but could not find a way to dell Gazu not to trigger video processing. Am I understanding this correctly?

**Reply by EmberLightVFX** [reply]
> * You can already do it via the Gazu client

Do you mean Gazu can upload a raw file that Kitsu can playback without any re-encoding? How would that code look like?

**Reply by flablog** [reply]
It's in the param `normalize_movie` which needs to be set to `False `in the `gazu.task.add_preview()` function if I'm correct:

https://github.com/cgwire/gazu/blob/d80485d911331012068879d05012f9b3a63726d7/gazu/task.py#L898C13-L898C13

EDIT: That will be using Gazu Python API

Using the restAPI, I guess you can add to the route `?normalize=false` 🤔 
https://github.com/cgwire/gazu/blob/d80485d911331012068879d05012f9b3a63726d7/gazu/task.py#L889C35-L889C35

**Reply by frankrousseau** [reply]
Yep, I don't remember why I put the parameter in the query but it is how it works.

**Comment by frankrousseau** [comment]
I add a link to a similar issue opened by @PeteDraper: https://github.com/cgwire/kitsu/issues/1312
